users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -201,5 +317,129 @@ def get_masked():
+    max_time_num_seq_len_divisor = tf.constant(config["max_time_num_seq_len_divisor"], dtype=tf.float32)
+    filter_factor = tf.cast(config["filter_factor"], tf.float32)
+    specaug_params = config["specaug_params"]
+
+    # Determine if we should use sorting
+    use_sorting = tf.logical_and(
+        config["enable_sorting"], tf.greater_equal(current_epoch, config["sorting_start_epoch"])
+    )
+
+    # Get filter indices (sorted or unsorted)
+    filter_layer = network.layers["features"].subnetwork.layers["conv_h_filter"].output.placeholder
+    num_filters = network.layers["features"].subnetwork.layers["conv_l"].output.shape[-1]
+
+    def get_sorted_indices():
+        sorted_filter_indices = sort_filters_by_center_freq(filter_layer)
+        sorted_indices = tf.stack(
+            [sorted_filter_indices * num_filters + filter_idx for filter_idx in range(num_filters)]
+        )
+        return tf.reshape(tf.transpose(sorted_indices), (-1,))
+
+    def get_unsorted_indices():
+        return tf.range(tf.shape(x)[data.feature_dim_axis])
+
+    sorted_indices = tf.cond(use_sorting, get_sorted_indices, get_unsorted_indices)
+
+    def get_masked():
+        x_masked = x
+        time_mask_max_size = tf.gather(specaug_params["time_mask_max_size"], current_epoch)
+        time_mask_max_num = tf.gather(specaug_params["time_mask_max_num"], current_epoch)
+        freq_mask_max_num = tf.gather(specaug_params["freq_mask_max_num"], current_epoch)
+        freq_mask_max_size = tf.gather(specaug_params["freq_mask_max_size"], current_epoch)
+        total_time_masks_max_frames = tf.cast(
+            tf.math.floor(config["time_mask_max_proportion"] * tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        total_freq_masks_max_size = tf.cast(
+            tf.math.floor(config["freq_mask_max_proportion"] * tf.cast(tf.shape(x)[data.feature_dim_axis], tf.float32)),
+            tf.int32,
+        )
+        max_time_num_seq_len = tf.cast(
+            tf.math.floordiv(
+                tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32),
+                tf.cast(1.0, tf.float32) / max_time_num_seq_len_divisor * tf.cast(time_mask_max_size, tf.float32),
+            ),
+            tf.int32,
+        )
+        # check for the limits
+        actual_time_mask_max_num = tf.minimum(
+            tf.maximum(
+                time_mask_max_num,
+                max_time_num_seq_len,
+            ),
+            total_time_masks_max_frames // time_mask_max_size,
+        )
+        actual_freq_mask_max_num = tf.minimum(freq_mask_max_num, total_freq_masks_max_size // freq_mask_max_size)
+
+        if config["filter_based_masking_strategy"] == "variance":
+            f_resp = get_frequency_response(filter_layer)
+            variance = tf.math.reduce_variance(f_resp, axis=0)
+            n_features = tf.shape(x)[data.feature_dim_axis]
+            probs = tf.nn.softmax(variance)
```

Comment:
Why do you take the softmax instead of just normalizing to unit sum?

Reply:
I think the properties of softmax fit well for this use case. If a filter stands out for a high variance it should be masked more often.

Reply:
Yes, it should be masked more often. But I'm afraid the softmax would flatten the distribution too much. You could do something like `probs = variance / sum(variance)` (pseudo-code). Could you please obtain the actual variance for true filters and compare the resulting `probs`?

Reply:
![image](https://github.com/user-attachments/assets/2879874f-b4ac-437b-87c0-e92830099023)
Softmax indeed looks like the wrong choice.